src/Elm/Parser/Declarations.elm

```diff
@@ -21,18 +23,38 @@ declaration : Parser State (Node Declaration)
 declaration =
     oneOf
         [ infixDeclaration
-        , function
-        , typeDefinition
         , portDeclaration
+        , maybeDocumentation
+            |> Combine.andThen
+                (\maybeDoc ->
+                    oneOf
+                        [ function maybeDoc
+                        , typeDefinition maybeDoc
+                        ]
+                )
         ]
 
 
-function : Parser State (Node Declaration)
-function =
+maybeDocumentation : Parser State (Maybe (Node Documentation))
+maybeDocumentation =
+    Comments.declarationDocumentation
+        |> Combine.ignore Layout.layoutStrict
+        |> maybe
+
+
+function : Maybe (Node Documentation) -> Parser State (Node Declaration)
+function maybeDoc =
     Node.parser functionName
         |> Combine.ignore (maybe Layout.layout)
         |> Combine.andThen functionWithNameNode
-        |> Combine.map (\f -> Node (Expression.functionRange f) (Declaration.FunctionDeclaration f))
+        |> Combine.map
+            (\f ->
+                Node
+                    { start = maybeDoc |> Maybe.map (Node.range >> .start) |> Maybe.withDefault (Expression.functionRange f).start
```

Comment:
Let's use a case expression instead of `Maybe.map+Maybe.withDefault`, that way we don't do compute `(Expression.functionRange f).start` unnecessarily.

Reply:
Actually, we already compute it for the `end` field, so let's extract it to a variable so that we only compute it once.